torch/_prims/rng_prims.py

```diff
@@ -132,5 +148,163 @@ def _philox_rand(
     )
 
 
+def register_run_and_save_rng_state_op():
+    run_and_save_rng_state = HigherOrderOperator("run_and_save_rng_state")
+
+    run_and_save_rng_state.fallthrough(DispatchKey.ADInplaceOrView)
+    run_and_save_rng_state.fallthrough(DispatchKey.PythonDispatcher)  # type: ignore[attr-defined]
+    run_and_save_rng_state.fallthrough(DispatchKey.PythonTLSSnapshot)  # type: ignore[attr-defined]
+
+    @run_and_save_rng_state.py_impl(DispatchKey.Autograd)
+    def impl_autograd(op, *args, **kwargs):
+        with torch._C._AutoDispatchBelowAutograd():
+            return run_and_save_rng_state(op, *args, **kwargs)
+
+    @run_and_save_rng_state.py_impl(DispatchKey.CUDA)
+    def impl_cuda(op, *args, **kwargs):
+        return torch.cuda.get_rng_state(), op(*args, **kwargs)
+
+    @run_and_save_rng_state.py_impl(DispatchKey.CPU)
+    def impl_cpu(op, *args, **kwargs):
+        return torch.get_rng_state(), op(*args, **kwargs)
+
+    @run_and_save_rng_state.py_impl(DispatchKey.BackendSelect)
+    def impl_backend_select(op, *args, **kwargs):
+        # with _ExcludeDispatchKeyGuard(DispatchKeySet(DispatchKey.BackendSelect)):
+        if kwargs.get("device"):
+            device = kwargs.get("device")
+            device_type = device if isinstance(device, str) else device.type  # type: ignore[union-attr]
+            if "cuda" in device_type:
+                impl = impl_cuda
+            elif "cpu" in device_type:
+                impl = impl_cpu
+            else:
+                raise NotImplementedError(
+                    f"Backend not available for device {device_type}"
+                )
+        else:
```

Comment:
Assuming there is a Tensor arg, we just re-dispatch and the dispatcher will take care of it for us:
```
with _ExcludeDispatchKeyGuard(DispatchKeySet(DispatchKey.BackendSelect)):
    run_and_save_rng_state(*args, **kwargs)
```

Reply:
I tried that earlier, but I was getting lots of failures. 

As of now, I get this error, when I use the above suggestion. It is going through my fake tensor impl, so something might be bad there?

~~~
  File "/scratch/anijain/work/pytorch/torch/_ops.py", line 252, in __call__
    return self.dispatch(dispatch_key_set.highestPriorityTypeId(), *args, **kwargs)
  File "/scratch/anijain/work/pytorch/torch/_ops.py", line 230, in dispatch
    return self.python_key_mode_table[type(curr_mode)](*args, **kwargs)
  File "/scratch/anijain/work/pytorch/torch/_prims/rng_prims.py", line 202, in impl_fake_tensor_mode
    return impl_backend_select(op, *args, **kwargs)
  File "/scratch/anijain/work/pytorch/torch/_prims/rng_prims.py", line 186, in impl_backend_select
    return run_and_save_rng_state(op, *args, **kwargs)
  File "/scratch/anijain/work/pytorch/torch/_ops.py", line 252, in __call__
    return self.dispatch(dispatch_key_set.highestPriorityTypeId(), *args, **kwargs)
  File "/scratch/anijain/work/pytorch/torch/_ops.py", line 230, in dispatch
    return self.python_key_mode_table[type(curr_mode)](*args, **kwargs)
  File "/scratch/anijain/work/pytorch/torch/_prims/rng_prims.py", line 202, in impl_fake_tensor_mode
    return impl_backend_select(op, *args, **kwargs)
  File "/scratch/anijain/work/pytorch/torch/_prims/rng_prims.py", line 186, in impl_backend_select
    return run_and_save_rng_state(op, *args, **kwargs)
  File "/scratch/anijain/work/pytorch/torch/_ops.py", line 245, in __call__
    flat_args = _to_flat_tuple(args, kwargs)
  File "/scratch/anijain/work/pytorch/torch/_ops.py", line 262, in _to_flat_tuple
    flat_args, _ = torch.utils._pytree.tree_flatten(args)
  File "/scratch/anijain/work/pytorch/torch/utils/_pytree.py", line 274, in tree_flatten
    if _is_leaf(pytree):
  File "/scratch/anijain/work/pytorch/torch/utils/_pytree.py", line 233, in _is_leaf
    return _get_node_type(pytree) not in SUPPORTED_NODES
  File "/scratch/anijain/work/pytorch/torch/utils/_pytree.py", line 227, in _get_node_type
    if _is_namedtuple_instance(pytree):
  File "/scratch/anijain/work/pytorch/torch/utils/_pytree.py", line 217, in _is_namedtuple_instance
    typ = type(pytree)
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
RecursionError: maximum recursion depth exceeded while calling a Python object

While executing %run_and_save_rng_state : [#users=2] = call_function[target=torch.ops.run_and_save_rng_state](args = (aten.native_dropout.default, %div_1, 0.1, True), kwargs = {})
Original traceback:
None

~~~

Reply:
Let me poke a bit at this locally and get back to you

Reply:
What is written in this function is not entirely faithful to what the PyTorch dispatcher does for regular operators, but I'm happy to accept this as-is to unblock.

The problem is that the PyTorch dispatcher uses "redispatch" to implement BackendSelect. What "redispatch" would do here is that for subsequent invocations of this operator, it masks out the dispatch keys above and including BackendSelect so that a dispatch key below it can be selected. Compare this to an ExcludeDispatchKeyGuard, which turns off the dispatch key for all operators (not just this operator!), which ends up causing problems down the line.

We have not implemented a "redispatch" for HigherOrderOperator yet.

Reply:
I see, that makes sense. I am totally willing to reimplement this function once we have redispatch for HigherOrderOp. So, lets go w/ the current design.